models/event.py

```diff
@@ -24,29 +29,106 @@
+  id = models.OneToOneField('auth.User', primary_key=True)
   credentials = CredentialsField()
   class Meta:
     app_label = 'tracker'
 
 
+class TimestampValidator(validators.RegexValidator):
+  regex = r'(?:(?:(\d+):)?(?:(\d+):))?(\d+)(?:\.(\d{1,3}))?$'
+  def __call__(self, value):
+    super(TimestampValidator, self).__call__(value)
+    h,m,s,ms = re.match(self.regex, unicode(value)).groups()
+    if h is not None and int(m) >= 60:
+      raise ValidationError('Minutes cannot be 60 or higher if the hour part is specified')
+    if m is not None and int(s) >= 60:
+      raise ValidationError('Seconds cannot be 60 or higher if the minute part is specified')
+
+
+class TimestampField(models.Field):
+  __metaclass__ = models.SubfieldBase
+  default_validators = [TimestampValidator()]
+  match_string = re.compile(r'(?:(?:(\d+):)?(?:(\d+):))?(\d+)(?:\.(\d+))?')
+
+  def __init__(self, always_show_h=False, always_show_m=False, always_show_ms=False, *args, **kwargs):
+    super(TimestampField, self).__init__(*args, **kwargs)
+    self.always_show_h = always_show_h
+    self.always_show_m = always_show_m
+    self.always_show_ms = always_show_ms
+  def to_python(self, value):
+    if isinstance(value, basestring):
+      try:
+        value = TimestampField.time_string_to_int(value)
+      except ValueError:
+        return value
+    if not value:
+      return 0
+    h,m,s,ms = value / 3600000, value / 60000 % 60, value / 1000 % 60, value % 1000
+    if h or self.always_show_h:
+      if ms or self.always_show_ms:
+        return '%d:%02d:%02d.%03d' % (h, m, s, ms)
+      else:
+        return '%d:%02d:%02d' % (h, m, s)
+    elif m or self.always_show_m:
+      if ms or self.always_show_ms:
+        return '%d:%02d.%03d' % (m, s, ms)
+      else:
+        return '%d:%02d' % (m, s)
+    else:
+      if ms or self.always_show_ms:
+        return '%d.%03d' % (s, ms)
+      else:
+        return '%d' % s
+  @staticmethod
+  def time_string_to_int(value):
+    try:
+      if str(int(value)) == value:
+        return int(value) * 1000
+    except ValueError:
+      pass
+    if not isinstance(value, basestring):
+      return value
+    if not value: return 0
+    match = TimestampField.match_string.match(value)
+    if not match:
+      raise ValueError('Not a valid timestamp: ' + value)
+    h,m,s,ms = match.groups()
+    s = int(s)
+    m = int(m or s / 60)
+    s %= 60
+    h = int(h or m / 60)
+    m %= 60
+    ms = int(ms or 0)
+    return h * 3600000 + m * 60000 + s * 1000 + ms
+  def pre_save(self, model, add):
+    return TimestampField.time_string_to_int(getattr(model, self.attname))
+  def get_internal_type(self):
+    return 'IntegerField'
+  def validate(self, value, model_instance):
+    super(TimestampField, self).validate(value, model_instance)
+    try:
+      TimestampField.time_string_to_int(value)
+    except ValueError:
+      raise ValidationError('Not a valid timestamp')
+
+
 def LatestEvent():
```

Comment:
This implementation of 'LatestEvent' differs from the one in models/donation.py. Why don't we just put a single, canonical implementation of it in __init__.py?